Optimize patmat analysis

retronym opened the pull request:
Avoid boxing/unboxing of the Lit value class by removing its
value-ness. Retain efficiency by caching a lit's negation and
caching its hashCode.

Avoid varargs penalty for creating 0-2 element clauses, and
use reusable instances for some sentinel array used in clause
simplification.

Use a List, rather than a Set, to track the assignments in
the current branch of the DPLL search.

Change the DPLL search to mutate the Array[Clause] in place
by inserting nulls for removed elements. We need to clone
the array when the search branches, but its a big improvement
over the old version of dropUnit.

Reuse a the pos/neg bitset during DPLL, clearing before use.
Switch to using a Java bitset which implements clear by emptying
the internal array rather than allocating a new one.

Hand roll the search for empty/unit clauses to avoid the overhead
of the ArrayOps combinators, which have to indirect array access
through Array_apply etc.

hrhino commented on src/reflect/scala/reflect/internal/util/IntArraySet.scala:
```diff
@@ -0,0 +1,98 @@
+    }
+  }
+  def iterator(): Iterator[Int] = {
+    data.slice(0, size).iterator
```
`data.iterator.slice(0, size)` to avoid copying?

retronym replied:
This isn't used anywhere too performance critical, but if so another option is:

```scala
  def intStream(): IntStream = {
    util.Arrays.stream(data, 0, size)
  }
```

retronym commented on src/reflect/scala/reflect/internal/util/IntArraySet.scala:
```diff
@@ -0,0 +1,98 @@
+
+import scala.util.hashing.MurmurHash3
+
+final class IntArraySet private (var data: Array[Int], var _size: Int) {
```
An alternative approach would be a pair `mutable.BitSet`-s

hrhino reviewed (commented):
I'll take a harder look over the weekend, but looks pretty great. I knew I should have just bit the mutable bullet last time I was in here.

It would be interesting to think about the use pattern of the set. If the literals are added out of order I wouldn't be surprised if the bitsets were faster.

retronym commented:
Running this and #6599 through JMH's GC profiler, with the `scala` corpus, and the [`patmat-exhaust-huge`](https://github.com/scala/compiler-benchmark/pull/66 corpus].

https://gist.github.com/80842a2cbadcb710ffc691859ed24f59

```
$ curl --silent -L https://gist.githubusercontent.com/retronym/80842a2cbadcb710ffc691859ed24f59/raw/1fd76aec453eb147116ac561325daddd37fdbcc5/gcprof.txt | egrep '((9e1d93ec81|b3cda4734e|ac5c8fa871 )|alloc.rate.norm +(latest|21d12e9))'
ac5c8fa871 (origin/2.13.x, 2.13.x) Merge pull request #6597 from retronym/faster/front-misc
[info] HotScalacBenchmark.compile:·gc.alloc.rate.norm                        latest                    false  2.13.0-pre-ac5c8fa-SNAPSHOT  patmat-exhaust-huge  sample    6  18179812077.333 ± 761927804.580    B/op
b3cda4734e (retronym/faster/patmat-subst, origin/pr/6599, faster/patmat-subst) Reduce allocation further in TypedSubstitution.<<
[info] HotScalacBenchmark.compile:·gc.alloc.rate.norm                        latest                    false  2.13.0-pre-b3cda47-SNAPSHOT  patmat-exhaust-huge  sample    6  18066069424.000 ± 771342941.017    B/op
9e1d93ec81 (retronym/faster/patmat-intset1, origin/pr/6598, faster/patmat-intset1) Avoid boxing in Set[Int] within patmat analysis
[info] HotScalacBenchmark.compile:·gc.alloc.rate.norm                        latest                    false  2.13.0-pre-9e1d93e-SNAPSHOT  patmat-exhaust-huge  sample    6  2475921313.778 ± 226008367.976    B/op
ac5c8fa871 (origin/2.13.x, 2.13.x) Merge pull request #6597 from retronym/faster/front-misc
[info] HotScalacBenchmark.compile:·gc.alloc.rate.norm                       21d12e9                    false  2.13.0-pre-ac5c8fa-SNAPSHOT     scala  sample    6  11505987592.000 ±   16643807.245    B/op
b3cda4734e (retronym/faster/patmat-subst, origin/pr/6599, faster/patmat-subst) Reduce allocation further in TypedSubstitution.<<
[info] HotScalacBenchmark.compile:·gc.alloc.rate.norm                       21d12e9                    false  2.13.0-pre-b3cda47-SNAPSHOT     scala  sample    6  11474755454.667 ±   34950424.793    B/op
9e1d93ec81 (retronym/faster/patmat-intset1, origin/pr/6598, faster/patmat-intset1) Avoid boxing in Set[Int] within patmat analysis
[info] HotScalacBenchmark.compile:·gc.alloc.rate.norm                       21d12e9                    false  2.13.0-pre-9e1d93e-SNAPSHOT     scala  sample    6  11918208816.000 ±   49044272.908    B/op
```

This PR reduces the reported allocation rate from 18.179GB/operation to 2.47GB/operation, which is the big win. Execution time also improves markedly. 

The numbers report that this PR slightly increases allocations in the `scala` corpus, which is a surprise and warrants investigation.

#6599 doesn't seem to have as much effect as I thought in either case.

I'll re-run those benchmarks with proper warmup and multiple forks to evaluate the effect on compilation time.

retronym commented:
Turns out that this PR is approximately neutral, allocation wise, for the non-pathological cases where patmat isn't a dominant factor. My comparison above was flawed: I'd was comparing this PR against a baseline (ac5c8fa871) that wasn't it its history.

I've added a commit that avoids a few extra allocations and answers a TODO from the first commit.

adriaanm commented on src/compiler/scala/tools/nsc/transform/patmat/Solving.scala:
```diff
@@ -124,6 +126,7 @@ trait Solving extends Logic {
 
       def isConst(l: Lit): Boolean = l == constTrue || l == constFalse
 
+      // The caller must release `clause`
```
Reformulate to explain why? "This method modifies `clause` in-place,..."

adriaanm commented on src/reflect/scala/reflect/internal/util/IntArraySet.scala:
```diff
@@ -0,0 +1,98 @@
+  }
+  def -=(i: Int): Unit = {
+    val index = binarySearch(i)
+    if (index == size - 1) {
```
Throw more appropriate exception when called on empty set? (size == 0, index == -1 --> `data(-1)` accessed)

adriaanm replied:
also, add comment that this is really two paths: fastpath for dropping last elem, and the illegal state path (empty array)

adriaanm commented on src/reflect/scala/reflect/internal/util/IntArraySet.scala:
```diff
@@ -0,0 +1,98 @@
+  }
+
+  override def equals(other: Any): Boolean = other match {
+    case ias: IntArraySet => ias.size == size && (size == 0 || util.Arrays.equals(ias.data, data))
```
add comment to explain use of `0` as removed element, which is fine because the sizes are checked first

adriaanm reviewed (commented):
A first round with some quick comments. I'll illustrate my key concern with a few pesky question :-)

Can you convince me that performance improvements merit the cost of maintaining IntArraySet? 

What if we added it to the stdlib? (would that cost still be acceptable? if not, why is it ok for the compiler?)

Is there a middle road? Could we use an IntMap? (EDIT: should IntMap provide overloads using Java's non-boxing SAM type?)

retronym commented:
I've pushed a commit to fix that corner case and expand the commentary.

I'm not wedded to this data structure. Usually I'd grab something from http://fastutil.di.unimi.it/docs/it/unimi/dsi/fastutil/ints/IntSet.html of the like, but I don't really want to introduce a dependency.

An alternative approach is to use a mutable bitset. This is appropriate if the values we're storing fall relatively densely within a smallish range (e.g. we wouldn't want to store `BitST(0, 1024, 16384)`). `Set[Lit]` might meet these requirements.

I think the main risk of `IntArraySet` here isn't the maintainence burden, but the chance that it is a poor data structure for some variants of the workload that pattern matcher throws at at. e.g Adding `is += 5; is += 4 is += 3 is += 2 is += 1 is += 0` involves a lot of copying.

non commented:
@retronym Any reason you chose to implement a set where you have to copy on most insertions? Did you consider a mutable, array-backed hash set? (For example, an implementation similar to [Debox](https://github.com/non/debox/blob/master/src/main/scala/debox/Set.scala), but manually specialized to `Int`?)

retronym commented:
Thinking on this some more, I now believe I've gone down the wrong path based on a flawed initial premise. I started by trying to remove the cost of boxed ints (the underlying of the `Lit` value class) in collections, which led me to trying `mutable.BitSet`, rolling my own `IntArraySet` and trying `IntHashSet` from debox.

That eventually led me to trying to shoehorn the same mutable int set into the backtracking DPLL algorithm, which incurred a cost to copy the sets to snapshot the state.

But perhaps the problem isn't the boxing itself, but rather the extra box/unbox that happens every time we interact with a `Lit` because it is defined as a value class. I'm now trying to de-value-classify it and restore to using a persistent/immutable Set.

retronym committed:
Optimize patmat analysis

Avoid boxing/unboxing of the Lit value class by removing its
value-ness. Retain efficiency by caching a lit's negation and
caching its hashCode.

Avoid varargs penalty for creating 0-2 element clauses, and
use reusable instances for some sentinel array used in clause
simplification.

Use a List, rather than a Set, to track the assignments in
the current branch of the DPLL search.

Change the DPLL search to mutate the Array[Clause] in place
by inserting nulls for removed elements. We need to clone
the array when the search branches, but its a big improvement
over the old version of dropUnit.

Reuse a the pos/neg bitset during DPLL, clearing before use.
Switch to using a Java bitset which implements clear by emptying
the internal array rather than allocating a new one.

Hand roll the search for empty/unit clauses to avoid the overhead
of the ArrayOps combinators, which have to indirect array access
through Array_apply etc.

retronym commented:
I've settled on a workable/performant solution, which currently sits in https://github.com/scala/scala/compare/2.13.x...retronym:faster/patmat-intset1-wip?expand=1

There are a few general collections performance improvements in there, which I added as I was shopping around for performance between `immutable.ChampHashSet`, `mutable.HashSet` and `immutable.TreeSet`.

I'll organize the branch and push the results here tomorrow.

retronym commented:
@adriaanm / @hrhino  This is ready for a fresh round of review. It is stacked on another PR that improves performance of ChampHashMap, so it's marked WIP until that one is merged.

hrhino commented on src/compiler/scala/tools/nsc/transform/patmat/Logic.scala:
```diff
@@ -224,19 +224,33 @@ trait Logic extends Debugging  {
         }
       }
 
+      def mapConverve[A <: AnyRef](s: Set[A])(f: A => A): Set[A] = {
```
"conserve"

retronym commented:
Some new benchmark results. These are for the baseline, the tip of this PR, and an additional commit that reverts the CHAMP optimizations we're discussing in #6725 

```
$ for v in e33743f 5e6029c badbd07; do (git --git-dir /code/scala/.git log --oneline -1 $v; cd /code/compiler-benchmark; sbt 'set scalaVersion in compilation := "2.13.0-pre-'$v'-SNAPSHOT"' 'hot -psource=patmat-exhaust-huge -f1 -wi 4 -i 4 -prof gc -o target/'$v.log); done
...
$ (for v in e33743f 5e6029c badbd07; do tail -n 18 /code/compiler-benchmark/compilation/target/$v.log; done ) | egrep '^(HotScalacBenchmark.compile|HotScalacBenchmark.compile:·gc.alloc.rate.norm)\s.*'

$ for v in e33743f 5e6029c badbd07; do (git --git-dir /code/scala/.git log --oneline -1 $v); done
e33743f31a (origin/2.13.x, 2.13.x) Merge pull request #6723 from viktorklang/wip-upgrade-jmh-√
5e6029c83c (retronym/faster/patmat-intset1) Optimize patmat analysis
badbd07ad9 (HEAD -> faster/patmat-intset1) Revert "Avoid allocating {Map,Set}Effect CHAMP in update/remove"

HotScalacBenchmark.compile                                            latest                    false  2.13.0-pre-e33743f-SNAPSHOT  patmat-exhaust-huge  sample    4        11437.867 ±       6778.199   ms/op
HotScalacBenchmark.compile:·gc.alloc.rate.norm                        latest                    false  2.13.0-pre-e33743f-SNAPSHOT  patmat-exhaust-huge  sample    4  21228478278.000 ± 4969165538.732    B/op
HotScalacBenchmark.compile                                            latest                    false  2.13.0-pre-5e6029c-SNAPSHOT  patmat-exhaust-huge  sample   12       3658.831 ±      290.268   ms/op
HotScalacBenchmark.compile:·gc.alloc.rate.norm                        latest                    false  2.13.0-pre-5e6029c-SNAPSHOT  patmat-exhaust-huge  sample    4  840821497.333 ± 24792967.684    B/op
HotScalacBenchmark.compile                                            latest                    false  2.13.0-pre-badbd07-SNAPSHOT  patmat-exhaust-huge  sample    9       5092.817 ±        412.756   ms/op
HotScalacBenchmark.compile:·gc.alloc.rate.norm                        latest                    false  2.13.0-pre-badbd07-SNAPSHOT  patmat-exhaust-huge  sample    4  856132183.333 ±   14336165.611    B/op
```

adriaanm commented:
3x faster and 25x less garbage? Nice!

retronym commented:
It's worth noting that these gains are only really relevant for programs with with patterns that stress the analysis, I'm using [patmat-exhaust-huge](https://github.com/scala/compiler-benchmark/blob/master/corpus/patmat-exhaust-huge/latest/test.scala) in the benchmark above.

retronym commented:
As expected, [benchmark results](https://scala-ci.typesafe.com/grafana/dashboard/db/pr-validation?orgId=1&from=1527950366093&to=1528269440394&var-scalaSha=5e6029c83c%7C71ac527296%7Ca9e7e35451%7C23b3701f58%7C84f3d57f40%7Cdd78188a6d%7C91a9d23324%7C1d8b8878d4%7Ce33743f31a%7C72d564f84f%7C2212d38a6f%7Cefe5393e09%7C004568c10d%7Cb8ed56f78e%7C4046983cd9%7C08ab3c427f%7C8b65ff4211%7C14da628cbf%7C8f004e876c%7C1478635d09%7C3961288bd7%7Cbbf0017573%7Cb6982c8dd9%7Ce4ba4579c0%7C8dfed294ea%7C596f047915%7C013b8c4b56%7C2007bf213a%7C21eb884c78%7Cf16eacca3d%7C5c5b5df67d%7C4e120ce786%7C15f427d5c4%7C6897b15677%7Ca6dd3a3ea0%7C0efb71066c%7C7f45554881%7C4e01e84e9c%7C00eb80ebab%7Cf5baf26ca4%7C6b988b1298%7C58157334d8%7C9b8b20ec87%7C3e19156ac3%7C9a8608c72c%7Cfd67f335c0%7Cd99e890c46%7C6a4b1f921d%7Cb7fef0e134%7C6b5c89a2c2%7Cc747ad56cf%7Cce5411eb38%7Cc8dddd010a%7C5d3847ebc6%7Ceef0c6a8df%7C6dd9fc7dc9%7Cb05f3b6123%7C23e74fa9a3%7Cbbf0c4fc93%7Cb787152428&var-source=All&var-bench=HotScalacBenchmark.compile&var-host=scalabench%40scalabench%40) are about flat on runtime for "regular" compilations:

retronym commented:
I've merged the CHAMP performance improvements to 2.13.x, rebased this atop it, and removed the WIP label.
